GIN guide (#52)
neuroinformatics-unit/HowTo: docs/source/open_science/GIN-repositories.md

## File locking

[File locking](https://gin.g-node.org/G-Node/Info/wiki/GIN+CLI+Usage+Tutorial#file-locking) is an important point in GIN and git-annex repositories which surprisingly comes up quite late in the GIN docs. Below, are the main ideas behind this.

- Files in a GIN repo can be _locked_ or _unlocked_.

- The lock state relates to the nature of the placeholder files we get in the working directory when we fetch the remote repository via `gin get <remote-repository-location>`:

  - **on Unix-like systems** (Mac, Ubuntu):
    - if a file is _locked_, its corresponding placeholder file will be a _symlink_. These symlinks point to the annexed content (under `.git/annex/objects`). With the symlinks we can open the file but not modify it.
    - If a file is _unlocked_, the placeholder file in the working directory is an _ASCII text file with a path_. The path is approximately where the content of the file will be downloaded to when we request it.

>   - **on Windows**:
>     - if a file is _locked_, the placeholder file is a _plain text file_ pointing to the content in the git annex.
>     - If a file is _unlocked_, the behaviour should be the same as in Unix-like systems (I haven't tested this directly).

Reviewer: ```suggestion
  - **on Windows**:
    - if a file is _locked_, the placeholder file is a _plain text file_ pointing to the content in the git annex(, but see caution below!).
    - If a file is _unlocked_, the behaviour is be the same as in Unix-like systems.
```

I confirm this is the case on Windows:
For a locked file:
```
$ file frames/10.08_10.08.2023-02-Right_frame_267960.png
frames/10.08_10.08.2023-02-Right_frame_267960.png: ASCII text, with no line terminators
```
After unlocking the file
```
$ file frames/10.08_10.08.2023-02-Right_frame_267960.png
frames/10.08_10.08.2023-02-Right_frame_267960.png: ASCII text
```

- The lock state of a file is _persistent_. This means that if I clone a GIN repo whose files are unlocked, I lock them in my local copy, and then upload that to the GIN server, the next time someone clones from the GIN repo the files they fetch will be locked.

- Unlocked files can be edited. If the data is unlocked and the full content of the dataset is downloaded locally, the file in the working directory has content, and so does its copy under git annex.

:::{caution}

> Notice that if we download locally the content of unlocked, the disk usage of the files checked into the repo doubles, because the content exists both in the working directory and under the git annex. But in exchange users can modify and revert files to previous commits.

Reviewer: ```suggestion
Note that if we download the contents of unlocked files locally, the disk usage of the files checked into the repo doubles, because the content exists both in the working directory and under the git annex. But in exchange users can modify and revert files to previous commits.
```
this sounds more natural to me, but I am not a native speaker.

> :::{caution}
> Notice that if we download locally the content of unlocked, the disk usage of the files checked into the repo doubles, because the content exists both in the working directory and under the git annex. But in exchange users can modify and revert files to previous commits.
> :::

Reviewer: ```suggestion
:::{caution}
Notice that if we download locally the content of unlocked, the disk usage of the files checked into the repo doubles, because the content exists both in the working directory and under the git annex. But in exchange users can modify and revert files to previous commits.
:::

:::{caution}
We have observed that it is possible to unintentionally overwrite locked files, at least on Windows. Please be careful and double-check the output of `gin ls` before uploading. You may also want to read [about `git-annex` on Windows](https://gin.g-node.org/G-Node/Info/wiki/Some+Notes+On+Git+Annex) if you are considering using `git-annex` directly.
:::
```

I was able to add a colourful triangle to a png file with Microsoft paint and commit it locally without any complaints.

Author: good news for Paint, identity crisis for locked files... 🙈 

thanks loads for testing and catching this btw!
